core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -846,4 +975,61 @@ public GroupByIdBlock getResult()
                             page.getPositionCount()));
         }
     }
+
+    private abstract class LowCardinalityDictionaryWork<T>
+            implements Work<T>
+    {
+        protected final Page page;
+        protected final int maxCardinality;
+        protected final int[] dictionarySizes;
+        protected final DictionaryBlock[] blocks;
+        protected int lastPosition;
+
+        public LowCardinalityDictionaryWork(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+            dictionarySizes = new int[channels.length];
+            blocks = new DictionaryBlock[channels.length];
+            int maxCardinality = 1;
+            for (int i = 0; i < channels.length; i++) {
+                Block block = page.getBlock(channels[i]);
+                verify(block instanceof DictionaryBlock, "Only dictionary blocks are supported");
+                blocks[i] = (DictionaryBlock) block;
+                int blockPositionCount = blocks[i].getDictionary().getPositionCount();
+                dictionarySizes[i] = blockPositionCount;
+                maxCardinality *= blockPositionCount;
+            }
+            this.maxCardinality = maxCardinality;
+        }
+
+        /**
+         * Returns combinations of all dictionaries ids for every position and populates
+         * samplePositions array with a single occurrence of every used combination
```

Comment:
Why do you need to pass `positionCount` as an argument? You already have `Page page` as a field? IMO composition or static method would be cleaner even though there might be a bit more code